src/bvh/optimization.rs

```diff
@@ -0,0 +1,359 @@
+
+// TODO Consider: Instead of getting the scene's shapes passed, let leaf nodes store an AABB
+// that is updated from the outside, perhaps by passing not only the indices of the changed
+// shapes, but also their new AABBs into optimize().
+
+impl BVH {
+    /// Optimizes the `BVH` by batch-reorganizing updated nodes.
+    /// Based on https://github.com/jeske/SimpleScene/blob/master/SimpleScene/Util/ssBVH/ssBVH.cs
+    ///
+    /// Needs all the scene's shapes, plus the indices of the shapes that were updated.
+    ///
+    pub fn optimize<Shape: BHShape>(&mut self,
+                                    refit_shape_indices: &HashSet<usize>,
+                                    shapes: &[Shape]) {
+        let mut refit_node_indices: HashSet<usize> = refit_shape_indices.iter()
+            .map(|x| shapes[*x].bh_node_index())
+            .collect();
+
+        // As long as we have refit nodes left, take the list of refit nodes
+        // with the highest depth (sweep nodes) and try to rotate them all
+        while refit_node_indices.len() > 0 {
+            let mut max_depth = 0;
+            let mut sweep_node_indices: Vec<usize> = Vec::new();
+
+            // Find max_depth and sweep_node_indices in one iteration
+            for refit_node_index in refit_node_indices.iter() {
+                let depth = match self.nodes[*refit_node_index] {
+                    BVHNode::Node { depth, .. } => depth,
+                    BVHNode::Leaf { depth, .. } => depth,
+                    BVHNode::Dummy => panic!("Dummy node found during BVH optimization!"),
+                };
+
+                if depth > max_depth {
+                    max_depth = depth;
+                    sweep_node_indices.clear();
+                    sweep_node_indices.push(*refit_node_index);
+                } else if depth == max_depth {
+                    sweep_node_indices.push(*refit_node_index);
+                }
+            }
+
+            // Try to find a useful tree rotation with all nodes previously found
+            for sweep_node_index in sweep_node_indices.iter() {
+                // This node does not need to be checked again
+                refit_node_indices.remove(&sweep_node_index);
+
+                let new_refit_node_index = self.try_rotate(*sweep_node_index, shapes);
+
+                // Instead of finding a useful tree rotation, we found another node
+                // that we should check, so we add its index to the refit_node_indices.
+                if let Some(index) = new_refit_node_index {
+                    refit_node_indices.insert(index);
+                }
+            }
+        }
+    }
+
+    /// Checks if there is a way to rotate a child and a grandchild (or two grandchildren) of
+    /// the given node (specified by `node_index`) that would improve the `BVH`.
+    /// If there is, the best rotation found is performed.
+    /// Relies on the children and transitive children of the given node having correct AABBs.
```

Comment:
`try_rotate` is not a public function anyways, so remarks about internal behaviour are okay.
This statement served as a reminder to

1. not unnecessarily check or update children AABBs
2. make sure to update the given node's AABBs before returning for later calls

I could remove the remark, of course.